app/models/program.rb

```diff
@@ -16,5 +16,58 @@ class Program < ApplicationRecord
 
   validates :facility_code, presence: true
   validates :program_type, inclusion: { in: InstitutionProgram::PROGRAM_TYPES }
-  validates_with ProgramValidator, on: :after_import
+
+  def self.after_import_batch_validations(_records, failed_instances, _row_offset)
+    duplicate_facility_description_results.each do |record|
+      message = line_number(record['csv_row']) + non_unique_error_msg(record['facility_code'], record['description'])
+      warning = { index: record['csv_row'], message: message }
+      failed_instances << warning
+    end
+
+    missing_facility_in_weam.each do |record|
+      program = Program.new(record)
+      message = line_number(program.csv_row) + BaseValidator.missing_facility_error_msg(program)
+      warning = { index: program.csv_row, message: message }
+      failed_instances << warning
+    end
+  end
+
+  def self.duplicate_facility_description_results
+    str = <<-SQL
+     SELECT programs.csv_row, programs.facility_code, programs.description
+      FROM programs
+        INNER JOIN (
+        SELECT
+          UPPER(facility_code) facility_code,
+          UPPER(description) description
+        FROM programs
+        GROUP BY
+          UPPER(facility_code),
+          UPPER(description)
+        HAVING COUNT(*) > 1
+        ) dupes on UPPER(programs.facility_code) = dupes.facility_code
+        AND UPPER(programs.description) = dupes.description
+    SQL
+
+    sql = Program.send(:sanitize_sql, [str])
+    Program.connection.execute(sql)
+  end
+
+  def self.missing_facility_in_weam
+    str = <<-SQL
+      SELECT programs.csv_row, programs.facility_code
+      FROM programs LEFT OUTER JOIN weams ON programs.facility_code = weams.facility_code
+      WHERE weams.facility_code IS NULL
+    SQL
+    sql = Program.send(:sanitize_sql, [str])
+    Program.connection.execute(sql)
```

Comment:
Can be replaced with:
```ruby
Program.joins('LEFT OUTER JOIN weams ON programs.facility_code = weams.facility_code').where(weams: { facility_code: nil})
```
Then you can use the objects above.

Reply:
changed